sphinx-build: Move code out of 'sphinx.__init__'

stephenfin opened the pull request:
Subject: sphinx-build: Move code out of `sphinx.__init__`

### Feature or Bugfix
- Refactor

### Purpose

We have multiple executables in tree and, while 'sphinx-build' is arguably the most important of these, there's no reason its importance should warrant inclusion at the package level.

Move this code to its own module, 'sphinx.build'. This name might be a bit disingenuous at present, given the availability of 'make-mode' here too, but that's an artifact of the current executable design and can be cleaned up later.

To avoid breaking packages that are using this feature directly, aliases for the old 'main' method are included. This is based on what [Django does](https://github.com/django/django/blob/1.11/django/test/runner.py#L688-L695) and, like Django, will allow us to safely remove the old modules in Sphinx 2.0.

### Detail
- N/A

### Relates
- N/A

stephenfin committed:
sphinx-build: Move code out of 'sphinx.__init__'

We have multiple executables in tree and, while 'sphinx-build' is
arguably the most important of these, there's no reason its importance
should warrant inclusion at the package level.

Create a new module, 'sphinx.cmd', and move the code from
'sphinx.__init__' into a 'build' submodule within. This name might be a
bit disingenuous at present, given the availability of 'make-mode' here
too, but that's an artifact of the current executable design and can be
cleaned up later.

To avoid breaking packages that are using this feature directly, aliases
for the old 'main' method are included. This is based on what Django
does [1] and, like Django, will allow us to safely remove the old
modules in Sphinx 2.0.

[1] https://github.com/django/django/blob/1.11/django/test/runner.py#L688-L695

Signed-off-by: Stephen Finucane <stephen@that.guru>

tk0miya commented:
@shimizukawa could you check this please.
I know you'd like to refactor command line interfaces on Sphinx-2.0.
I'm afraid this conflicts with the idea.

stephenfin commented:
> @shimizukawa could you check this please.
> I know you'd like to refactor command line interfaces on Sphinx-2.0.
> I'm afraid this conflicts with the idea.

I'd like a little more info on this too. I've addressed #3259 locally, but have not pushed it because I'm waiting on #3929 to merge. Would be good to verify that I haven't wasted my time on this :smile:

tk0miya commented:
I heard that he would like to restruct command line interface of Sphinx.
His idea is all commands (`sphinx-*`) will be integrated to sub-commands of `sphinx` command like following:
```
$ sphinx quickstart   (sphinx init?)
$ sphinx build
$ sphinx apidoc
# and so on.
```

Of course, we should keep current commands to keep compatibility for a while.
I don't know the change is important or not, but I feel good.

stephenfin commented:
> I heard that he would like to restruct command line interface of Sphinx.
> His idea is all commands (sphinx-*) will be integrated to sub-commands of sphinx command like following:
> ```
> $ sphinx quickstart   (sphinx init?)
> $ sphinx build
> $ sphinx apidoc
> # and so on.
> ```

Ah, that makes sense. FWIW, it would be good to look at something like [`cliff`](https://docs.openstack.org/cliff/latest/) for this, which would allow plugins for folks that want to add other `sphinx` sub-commands. I'd be happy to look at this is @shimizukawa hadn't done anything yet.
_However_, while we definitely should wait for @shimizukawa's input, I can't see how this would have any ill-effects for that goal. If anything, it would make it easier, as it would allow us to include place the `sphinx` command stuff in `__init__.py` and call into `sphinx.build` for sub-commands.

shimizukawa commented:
Actually, I have not done anything, so this change will not conflict with any ideas (I have not thought of anything yet!)

shimizukawa reviewed (approved):
LGTM!

stephenfin commented on sphinx/__init__.py:
```diff
@@ -63,62 +60,37 @@
+        RemovedInSphinx20Warning,
+        stacklevel=2,
+    )
+    build.make_main(*args, **kwargs)
```
Oops - this should be `build_main`. I'll rework

stephenfin commented:
I fixed a typo. The patch is otherwise unchanged.

tk0miya commented:
OMG, I noticed `sphinx/build.py` prevents my development. It blocks bash-completion for `sphinx/builders/`. I feel that is big problem to me. So I'd like to rename it to another naming...
How about `sphinx.cmdline.build` or `sphinx.commands.build`?

stephenfin committed:
Stop handling package issues from 'sphinx-build'

There were a number of package error handlers run as part of the
'sphinx-build' command/executable:

- Unsupported Python version (it should be 2.7 or 3.4+)
- Missing packages (missing docutils, jinja2, and roman, which is part
  of docutils, packages)
- Out-of-date packages (docutils)

This code is mostly unchanged since Sphinx was first released. Python,
and in particular Python's packaging options, have come a long way since
then. Today, all of the above checks are provided by setuptools and the
'setup.py' script, meaning we should never actually get to the point of
triggering any of these checks. This is further reinforced by the fact
that none of the other executables carry out these checks: either this
is a bug that no one has reported in ~8 years or, more likely, the
checks are useless and we don't need them anywhere.

In all, we can happily remove these checks, greatly simplify a piece of
code that's otherwise rarely touched, and trust that setuptools is up to
the job it's designed for.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin commented:
I'll do that now. Spotted some other cleanups that I could do in the interim

stephenfin committed:
sphinx-quickstart: Move code to 'sphinx.cmd'

We're going to move the executable's here (or at least those that part
of the core library). The 'sphinx-build' executable was already moved,
so lets do 'sphinx-quickstart' next.

To avoid breaking packages that are using this feature directly, aliases
for the old 'main' method are included. This is based on what Django
does [1] and, like Django, will allow us to safely remove the old
modules in Sphinx 2.0.

[1] https://github.com/django/django/blob/1.11/django/test/runner.py#L688-L695

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin commented:
OK, I've moved this to `sphinx.cmd.build` instead. I haven't moved `sphinx.cmdline` and `sphinx.make_mode`, even though I wanted to, because I feel make mode could do with some work first. I'm going to wait for #4097 to merge first before doing any major surgery on this.

tk0miya commented:
Thanks!
